File: cracking_coding_interview/chap_4/4.3_list_of_depths.rb
class ListNode
  def initialize(val)
    @val = val
    @next = nil
  end
end

def list_of_depths(root)
  return [] if !root
  return [ListNode.new(root.val)] if !root.left && !root.right

  queue = [root]
  result = []

  while queue.length > 0
    sub_arr = []
    length = queue.length

    (0...length).each do |idx|
      cur_node = queue.shift
      new_node = ListNode.new(cur_node.val)

      sub_arr[idx-1].next = new_node if idx != 0
      sub_arr << new_node

      queue << cur_node.left if cur_node.left
      queue << cur_node.right if cur_node.right
    end
    result << sub_arr[0]
  end
  result
end

File: cracking_coding_interview/chap_4/4.3_list_of_depths.js
class ListNode {
    constructor(val) {
    this.val  = val;
    this.next = null;
    }
}

const listOfDepths = root => {
    if (!root) return [];
    if (!root.left && !root.right) return [new ListNode(root.val)];

    const queue = [root],
          result = [];

    while (queue.length > 0) {
        const subArr = [],
              length = queue.length;
        for (let i = 0; i < length; i++) {
            const curNode = queue.shift(),
                  newNode = new ListNode(curNode.val);
            if (i !== 0) {
                subArr[i-1].next = newNode;
            }
            subArr.push(newNode);
            if (curNode.left) queue.push(curNode.left);
            if (curNode.right) queue.push(curNode.right);
        }
        result.push(subArr[0]);
    }
    return result;

};
